test/test_pebble.py

```diff
@@ -2617,14 +2660,16 @@ def setUp(self):
         time_patcher.start()
         self.addCleanup(time_patcher.stop)
 
-    def add_responses(self, change_id, exit_code, change_err=None):
+    def add_responses(self, change_id: str, exit_code: int,
+                      change_err: typing.Optional[str] = None):
```

Comment:
Curious that this doesn't need a return type annotation -- does Pyright infer that?

Reply:
Yes. In general, it seems that return types are not required unless you want the type to be wider/narrower than what it's actually returning with the current code - or if you're doing it for documentation purposes rather than for the type checker.

It seems like what pyright does is construct an implied return type that's the union of the types of whatever gets returned by the function, so as long as it knows those types, it gets the rest. Here, it knows that `stdin`, `stdout`, and `stderr` are `MockWebSocket` objects, so 'figures out' that the return type is `Tuple[MockWebSocket, MockWebSocket, MockWebSocket]`, since that's the only return statement. Or if there's no `return` statement, then it assumes `-> None`.

Reply:
Interesting. From a while back I remember that MyPy was different in this respect -- it requires return type annotations. Or maybe it was that functions like `def foo():` weren't type checked under MyPy (but they are under Pyright), so you'd have to explicitly declare it as `def foo() -> None:` to get MyPy to type check it. I added some `-> None`s to methods/functions that took no parameters a while back.

Reply:
I suspect maybe this is something that's changed over time (pyright certainly seems to pump out releases full of changes at a furious rate). Right now both seem ok with it:

```
tameyer@tam-canoncial-1:~/code/operator$ cat /tmp/returntypes.py 
def a():
    return "hello world!"

def b(i: int):
    return f"got {i}"

def c(i: int):
    if i > 0:
        return 'string'
    return b'bytes'

def d():
    a()
tameyer@tam-canoncial-1:~/code/operator$ pyright /tmp/returntypes.py 
0 errors, 0 warnings, 0 informations 
tameyer@tam-canoncial-1:~/code/operator$ mypy /tmp/returntypes.py 
Success: no issues found in 1 source file
```